unixfs/src/file.rs

```diff
@@ -0,0 +1,479 @@
+
+            let mh = multihash::wrap(multihash::Code::Sha2_256, &result[..]);
+            let cid = Cid::new_v0(mh).unwrap();
+
+            match self.blocks.insert(cid.clone(), block.to_vec()) {
+                None => {}
+                _ => unreachable!("duplicate cid {}", cid),
+            }
+
+            cid
+        }
+
+        fn with_fixtures() -> Self {
+            let mut this = Self::default();
+            let foobar_blocks: &[&[u8]] = &[
+                // root for "foobar\n" from go-ipfs 0.5 add -s size-2
+                //     root
+                //      |
+                //  ----+-----
+                //  |  |  |  |
+                // fo ob ar  \n
+                // QmRJHYTNvC3hmd9gJQARxLR1QMEincccBV53bBw524yyq6
+                &hex!("12280a221220fef9fe1804942b35e19e145a03f9c9d5ca9c997dda0a9416f3f515a52f1b3ce11200180a12280a221220dfb94b75acb208fd4873d84872af58bd65c731770a7d4c0deeb4088e87390bfe1200180a12280a221220054497ae4e89812c83276a48e3e679013a788b7c0eb02712df15095c02d6cd2c1200180a12280a221220cc332ceb37dea7d3d7c00d1393117638d3ed963575836c6d44a24951e444cf5d120018090a0c080218072002200220022001"),
+                // first bytes: fo
+                &hex!("0a0808021202666f1802"),
+                // ob
+                &hex!("0a08080212026f621802"),
+                // ar
+                &hex!("0a080802120261721802"),
+                // \n
+                &hex!("0a07080212010a1801"),
+
+                // same "foobar\n" but with go-ipfs 0.5 add --trickle -s size-2
+                &hex!("12280a2212200f20a024ce0152161bc23e7234573374dfc3999143deaebf9b07b9c67318f9bd1200180a12280a221220b424253c25b5a7345fc7945732e363a12a790341b7c2d758516bbad5bbaab4461200180a12280a221220b7ab6350c604a885be9bd72d833f026b1915d11abe7e8dda5d0bca689342b7411200180a12280a221220a8a826652c2a3e93a751456e71139df086a1fedfd3bd9f232ad52ea1d813720e120018090a0c080218072002200220022001"),
+                // the blocks have type raw instead of file, for some unknown reason
+                &hex!("0a0808001202666f1802"),
+                &hex!("0a08080012026f621802"),
+                &hex!("0a080800120261721802"),
+                &hex!("0a07080012010a1801"),
+            ];
+
+            for block in foobar_blocks {
+                this.insert_v0(block);
+            }
+
+            this
+        }
+    }
+
+    #[test]
+    fn balanced_traversal() {
+        let target = "QmRJHYTNvC3hmd9gJQARxLR1QMEincccBV53bBw524yyq6";
+        let blocks = FakeBlockstore::with_fixtures();
+
+        // filled on root
+        let (mut links_and_ranges, mut state) = {
+            let root = FileReader::from_block(blocks.get_by_str(target)).unwrap();
+
+            let (mut links_and_ranges, traversal) = match root.content() {
+                (FileContent::Spread(iter), traversal) => (
+                    iter.map(|(link, range)| (link.Hash.unwrap_borrowed().to_vec(), range))
+                        .collect::<Vec<(Vec<u8>, _)>>(),
```

Comment:
I _think_ this one could remain an iterator (i.e. doesn't require a `collect`) - it seems to only be traversed once in reverse, which can be done with `.rev()`